src/rt/sections_elf_shared.d

```diff
@@ -711,12 +711,12 @@ void scanSegments(in ref dl_phdr_info info, DSO* pdso)
         case PT_LOAD:
             if (phdr.p_flags & PF_W) // writeable data segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._gcRanges.insertBack(beg[0 .. phdr.p_memsz]);
             }
             version (Shared) if (phdr.p_flags & PF_X) // code segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._codeSegments.insertBack(beg[0 .. phdr.p_memsz]);
```

Comment:
> Agreed. However, the fix to align the _codeSegments[0] address rounds down. So theoretically this could already cause a out of bounds read.

Shouldn't be, as it's still within range of `[dlpi_addr .. p_vaddr]` unless `dlpi_addr` is unaligned, and `p_vaddr` is 0.  Which I don't think it can ever be if `(p_vaddr & p_align) == (p_offset & p_align)` is to be satisfied.

> So I guess the GC simply casts the _codeSegments to size_t[] as indicated by your example?

This is simple pointer arithmetic, you increment a pointer, its address with be incremented by `size_t.sizeof`.
```
void *p = 0x80;
p++; // p == 0x84 or 0x88
p++; // p == 0x88 or 0x90
```
And as all pointers in the data section should be guaranteed aligned to `size_t.sizeof`, you should make sure that the start address is aligned on the same boundary, otherwise the GC will be scanning for pointers at an unaligned offset.  My assumption is that in this case, runtime will find no pointers to GC memory.